module/zfs/zap.c

```diff
@@ -1378,6 +1508,245 @@ fzap_get_stats(zap_t *zap, zap_stats_t *zs)
+		lastblk = zap_f_phys(zap)->zap_ptrtbl.zt_blk +
+		    zap_f_phys(zap)->zap_ptrtbl.zt_numblks - 1;
+	} else {
+		/* Embedded ptrtbl */
+		nentries = (1 << ZAP_EMBEDDED_PTRTBL_SHIFT(zap));
+		lastblk = 0;
+	}
+
+	for (uint64_t idx = 0; idx < nentries; idx++) {
+		uint64_t blk;
+		zap_idx_to_blk(zap, idx, &blk);
+		if (blk > lastblk)
+			lastblk = blk;
+	}
+
+	ASSERT3U(lastblk, <, zap_f_phys(zap)->zap_freeblk);
+
+	zap_f_phys(zap)->zap_freeblk = lastblk + 1;
+}
+
+/*
+ * ZAP shrinking algorithm.
+ *
+ * We shrink ZAP recuresively removing empty leaves. We can remove an empty leaf
+ * only if it has a sibling. Sibling leaves have the same prefix length and
+ * their prefixes differ only by the least significant (sibling) bit. We require
+ * both siblings to be empty. This eliminates a need to rehash the non-empty
+ * remaining leaf. When we have removed one of two empty sibling, we set ptrtbl
+ * entries of the removed leaf to point out to the remaining leaf. Prefix length
+ * of the remaining leaf is decremented. As a result, it has a new prefix and it
+ * might have a new sibling. So, we repeat the process.
+ *
+ * Steps:
+ * 1. Check if a sibling leaf (sl) exists and it is empty.
+ * 2. Release the leaf (l) if it has the sibling bit (slbit) equal to 1.
+ * 3. Release the sibling (sl) to derefer it again with WRITER lock.
+ * 4. Upgrade zapdir lock to WRITER (once).
+ * 5. Derefer released leaves again.
+ * 6. If it is needed, recheck whether both leaves are still siblings and empty.
+ * 7. Set ptrtbl pointers of the removed leaf (slbit 1) to point out to blkid of
+ * the remaining leaf (slbit 0).
+ * 8. Free disk block of the removed leaf (dmu_free_range).
+ * 9. Decrement prefix_len of the remaining leaf.
+ * 10. Repeat the steps.
+ */
+static int
+zap_shrink(zap_name_t *zn, zap_leaf_t *l, dmu_tx_t *tx)
+{
+	zap_t *zap = zn->zn_zap;
+	int64_t zt_shift = zap_f_phys(zap)->zap_ptrtbl.zt_shift;
+	uint64_t hash = zn->zn_hash;
+	uint64_t prefix = zap_leaf_phys(l)->l_hdr.lh_prefix;
+	uint64_t prefix_len = zap_leaf_phys(l)->l_hdr.lh_prefix_len;
+	boolean_t trunc = 0;
+	int nshrunk = 0;
+	int err = 0;
+
+	ASSERT3U(zap_leaf_phys(l)->l_hdr.lh_nentries, ==, 0);
+	ASSERT3U(prefix_len, <=, zap_f_phys(zap)->zap_ptrtbl.zt_shift);
+	ASSERT(RW_LOCK_HELD(&zap->zap_rwlock));
+	ASSERT3U(ZAP_HASH_IDX(hash, prefix_len), ==, prefix);
+
+	boolean_t writer = B_FALSE;
+
+	/*
+	 * To avoid deadlock always deref leaves in the same order -
+	 * sibling 0 first, then sibling 1.
+	 */
+	while (prefix_len) {
+		zap_leaf_t *sl;
+		int64_t prefix_diff = zt_shift - prefix_len;
+		uint64_t sl_prefix = prefix ^ 1;
+		uint64_t sl_hash = ZAP_PREFIX_HASH(sl_prefix, prefix_len);
+		int slbit = prefix & 1;
+
+		ASSERT3U(zt_shift, ==, zap_f_phys(zap)->zap_ptrtbl.zt_shift);
+		ASSERT3S(zap_leaf_phys(l)->l_hdr.lh_nentries, ==, 0);
+
+
+		/*
+		 * Check if there is a sibling by reading prttbl ptrs.
+		 */
+		if (zap_check_leaf_by_ptrtbl(zap, sl_prefix, prefix_len) == 0)
+			break;
```

Comment:
I can't imagine why do we need this check here unless we want to check for hash table corruption.  We got here while holding both ZAP and leaf locks, so unless we have corruption, this must always be true, since we just got prefix and length from the leaf, and leaf prefix length must always be smaller or equal to hash prefix length, and the hash table pointers should be set accordingly.  There can be no ongoing simultaneous shrinking in other threads while we are holding the locks, since it would require the ZAP lock.

Same time, after we drop the ZAP  lock on upgrade failure we can not trust anything we know about the ZAP.  Now there is no way to convert FatZAP into MicroZAP IIRC, otherwise even that could happen.  All we have at that point is prefix and length, that may or may not represent any particular leafs, and that is where IMO the check has to be done.

Reply:
I don't quite get your point.
This function is optimization. It is fast. It checks whether a sibling leaf exists looking into the zap ptrtbl. So, if there is no 
a sibling leaf, we just don't have to call zap_deref_leaf().

Reply:
I'm sorry, but I do not understand what this function optimizes here.  As I understand, the sibling leaf should always exist at this point.  Current leaf prefix can be smaller than hash prefix, i.e. single leaf can be referenced by multiple hash entries within its prefix, but it should never be referenced by hash entries outside of its prefix.  I was thinking that we may need this to be re-checked sanity after we dropped/reacquired the lock, but looking now you are already rechecking `lh_prefix_len != prefix_len` for both leaves, that should be sufficient to protect from simultaneous shrinking from two sides.  So please explain me what does this optimize and what happen if we just remove this function call?

Reply:
No, at this point we don't know whether the sibling leaf exists or not. There are 2 ways to find out:

1) zap_deref_leaf() - we "dereference" a leaf with prefix/prefix_len corresponding to the sibling leaf. The leaf we got might or might not be the sibling - we have to figure that out checking its prefix/prefix_len.

2) We can say whether the sibling leaf exists just looking into ptrtbl, calling zap_check_leaf_by_ptrtbl().

If we remove zap_check_leaf_by_ptrtbl(), then we always have one additional zap_deref_leaf() call.
zap_deref_leaf() might have to read a leaf block from disk, and it is much more heavy than zap_check_leaf_by_ptrtbl().

Reply:
OK. I'm sorry, I got my mistake.  I thought you are calling zap_check_leaf_by_ptrtbl() for the current leaf to make it also check for the sibling, but you are calling it for the sibling prefix.  This way it makes more sense, checking first and last hash entries for expected sibling to guess its prefix that way.  What I don't understand then is why in zap_check_leaf_by_ptrtbl() you need to check for sibling of that sibling.  Since sibling of the sibling is the current leaf, it should always exist.  Shouldn't it?  It would reduce zap_check_leaf_by_ptrtbl() in half.

Reply:
When we call zap_check_leaf_by_ptrtbl(), the sibling range of entries is known and limited. It is limited by the range of entries of the current leaf. In such case, we don't need the last two conditions. But, if we don't have a "current" leaf, we cannot imply anything about the range. So, we have to check entries before-first, after-last.
Because, at the moment, zap_shrink() is the only consumer of zap_check_leaf_by_ptrtbl(), I don't mind to remove those two last conditions.

Reply:
It would not confuse me so much if zap_check_leaf_by_ptrtbl() don't do unneeded things.  So I vote for simplicity.  Also extra hash table lookup is far from free if zap_ptrtbl.zt_numblks != 0.  If the leaf is in cache, I am not even convinced that this optimization is good, though the hash table indeed should be in case much more often, so whatever.

Reply:
I renamed the function because with the applied changes it always require an existing initial sibling to exist

Reply:
zap_check_leaf_by_ptrtbl() was a general-purpose function - we detect if there was a leaf with sl_prefix/prefix_len for all cases.
Now, we just cannot detect that unless we already have an initial leaf (a sibling of the sibling).
I think, the word "sibling" just says that check_sibling_ptrtbl_range() is a special-purpose function.
Does it make sense?

Reply:
check_sibling_ptrtbl_range() now knows nothing about any siblings.  When given some arbitrary prefix, it validates the pointers.  The fact that it is used for siblings now does not affect its functioning in any way, so I don't see a point to mention it in the name or the comment.

Reply:
When we call the function we have a leaf with prefix/prefix_len. Because of that, we know for sure, that there cannot exist a bigger range in ptrtbl than the range corresponding to sl_prefix/prefix_len, right? That why we can remove those additional "ifs". Otherwise, if there is no initial leaf with prefix/prefix_len we must check entries "before_first" and "after_last", because we can hit inside a bigger range, which corresponds to some leaf with smaller prefix_len.
For example: there is only 1 leaf - prefix_len == 0, and we try to find if there is leaf with some prefix_len > 0. The function will return a false positive value.

Reply:
Yes, function checks only for leafs smaller than expected, not bigger.  You may write that in a comment, describing what function does and does not.  It still does not explain why function should be called "sibling" if it has nothing to do about siblings.  Theoretically we may know maximum size of the leaf from some other source. 

Whatever, I am tired of this bikeshedding. I'll leave it to you, but my opinion I've shared.